test/provider/AppDirectory.unittest.ts

```diff
@@ -0,0 +1,109 @@
+import 'jest';
+import 'reflect-metadata';
+
+import {AppDirectory} from '../../src/provider/model/AppDirectory';
+import {Application} from '../../src/client/directory';
+
+
+enum StorageKeys {
+    URL = 'url',
+    APPLICATIONS = 'applications'
+}
+
+const appTemplate: Application = {
+    appId: 'id',
+    name: 'test-app',
+    manifest: '',
+    manifestType: '',
+    intents: [
+        {
+            name: 'Call',
+            contexts: [
+                'dial'
+            ],
+            customConfig: {}
+        }
+    ]
+};
+
+// Generate applications
+const generateApplications = (amount: number): Application[] => {
+    return Array(amount).fill(appTemplate).map((app, i) => ({
+        ...app,
+        appId: `id-${i}`,
+        name: `${app.name}-${i}`
+    }));
+};
+
+describe('AppDirectory Unit Tests', () => {
+    let appDirectory: AppDirectory;
+    let store: LocalForage;
+
+    afterEach(async () => {
+        await store.clear();
+    });
+
+    describe('Storage', () => {
+        beforeEach(async () => {
+            appDirectory = new AppDirectory();
+        });
+
+        afterEach(async () => {
+            await store.clear();
+        });
+
+        it('Correctly sets up the store', async () => {
+            store = appDirectory['_store'];
+            // await store.ready();
+            const url = await store.getItem(StorageKeys.URL);
+            const applications = await store.getItem(StorageKeys.APPLICATIONS);
+            expect(url).toEqual(null);
+            expect(applications).toEqual(null);
+        });
+
+        describe('Storing fetchData', () => {
+            it('Correctly sets applications in the store', async () => {
+                appDirectory['fetchData'] = jest.fn(() => Promise.resolve([]));
+                await appDirectory.getAllApps();
+                const applications = await store.getItem(StorageKeys.APPLICATIONS);
+                expect(applications).toEqual([]);
+            });
+
+            it('Correctly stores fetchData in the store and in memory', async () => {
+                const fakeApps = generateApplications(3);
+                appDirectory['fetchData'] = jest.fn(() => Promise.resolve(fakeApps));
```

Comment:
This is a potentially risky way to test the fetch functionality. We're testing that the directory correctly fetches app data, but also replacing the code within the directory that performs that fetch.

This kind of mocking/stubbing should really only be used on things outside of the class being tested - so for example, leave the class completely unmodified, but mock the `fetch` method - since this will change the behaviour of the application without modifying the actual code that's being tested.

Of course, replacing a global such as `fetch` is also potentially risky, as it would leak across tests if not careful. Make sure the "real" `fetch` function is saved first and restored afterward - we could do with some test utils here.